src/tests/ftest/launch.py

```diff
@@ -20,164 +20,550 @@
+            daos_files[key] = value
+
+    return daos_files
+
+
+def get_hosts_from_yaml(test_yaml):
+    """Extract the list of hosts from the test yaml file.
+
+    This host will be included in the list if no clients are explicitly called
+    out in the test's yaml file.
+
+    Args:
+        test_yaml (str): test yaml file
+
+    Returns:
+        list: a unique list of hosts specified in the test's yaml file
+
+    """
+    yaml_data = get_yaml_data(test_yaml)
+    server_keys = [
+        "test_machines",
+        "test_servers",
+        "daos_servers",
+        "servers",
+        "test_machines1",
+        "test_machines2",
+        "test_machines2a",
+        "test_machines3",
+        "test_machines6"]
+    client_keys = [
+        "test_clients",
+        "clients",
+    ]
+    host_set = set()
+    found_client_key = False
+    matches = find_values(yaml_data, server_keys + client_keys)
+    for key, value in matches.items():
+        host_set.update(value)
+        if key in client_keys:
+            found_client_key = True
+
+    # Include this host as a client if no clients are specified
+    if not found_client_key:
+        host_set.add(socket.gethostname().split(".")[0])
+
+    return sorted(list(host_set))
+
+
+def clean_logs(test_yaml):
+    """Remove the test log files on each test host.
+
+    Args:
+        test_yaml (str): yaml file containing host names
+    """
+    # Use the default server yaml and then the test yaml to update the default
+    # DAOS log file locations.  This should simulate how the test defines which
+    # log files it will use when it is run.
+    log_files = get_log_files(test_yaml, get_log_files(BASE_LOG_FILE_YAML))
+    host_list = get_hosts_from_yaml(test_yaml)
+    command = "ssh {{host}} \"rm -fr {}\"".format(" ".join(log_files.values()))
+    print("Cleaning logs on {}".format(host_list))
+    spawn_commands(host_list, command)
+
+
+def archive_logs(avocado_logs_dir, test_yaml):
+    """Copy all of the host test log files to the avocado results directory.
+
+    Args:
+        avocado_logs_dir ([type]): [description]
+        test_yaml (str): yaml file containing host names
+    """
+    this_host = socket.gethostname().split(".")[0]
+    log_files = get_log_files(TEST_LOG_FILE_YAML)
+    host_list = get_hosts_from_yaml(test_yaml)
+    doas_logs_dir = os.path.join(avocado_logs_dir, "latest", "daos_logs")
+
+    # Create a subdirectory in the avocado logs directory for this test
+    print("Archiving host logs from {} in {}".format(host_list, doas_logs_dir))
+    get_output("mkdir {}".format(doas_logs_dir))
+
+    # Create a list of log files that are not directories
+    non_dir_files = [
+        log_file for log_file in log_files.values()
+        if os.path.splitext(os.path.basename(log_file))[1] != ""]
+
+    # Copy any log files that exist on the test hosts and remove them from the
+    # test host if the copy is successful
+    command = "set -e; ssh {{host}} '{}'".format(
```

Comment:
Here, you want the `set -e` in the shell script run on the remote.  Since `command` is just the single `ssh` command, no need to `set -e` for it:
```suggestion
    command = "ssh {{host}} 'set -e; {}'".format(
```

Or just move the `set -e` into the script string below.

The idea is that any error run in that (remote) shell script should exit the script with an error code, which will bubble up to `ssh` exiting with an error code so that the caller knows it failed (although at the moment, I guess that gets lost in `spawn_commands` -- a future enhancement).  An argument could be made for adding error checking to the script, allowing it to continue past errors but still exit with an error if one was encountered in the processing, but I think that can be a future modification.

So the only real action here is to move the `set -e` into the remote shell script instead of the local one.

Reply:
Moving `set -e` as you suggested does yield an error in the `launch.py` output:
```
Archiving host logs from ['boro-41', 'boro-43', 'boro-52', 'boro-53'] in /home/pahender/avocado/job-results/latest/daos_logs
Running mkdir /home/pahender/avocado/job-results/latest/daos_logs
Running ssh boro-41 'set -e; for file in /tmp/pahender/*_my_daos.log /tmp/pahender/my_daos.log /tmp/daos_agent.log /tmp/daos_control.log /tmp/test_server.log; do if [ -e $file ]; then scp $file boro-41:/home/pahender/avocado/job-results/latest/daos_logs/${file##*/}-boro-41 && rm -fr $file; fi; done'
Running ssh boro-43 'set -e; for file in /tmp/pahender/*_my_daos.log /tmp/pahender/my_daos.log /tmp/daos_agent.log /tmp/daos_control.log /tmp/test_server.log; do if [ -e $file ]; then scp $file boro-41:/home/pahender/avocado/job-results/latest/daos_logs/${file##*/}-boro-43 && rm -fr $file; fi; done'
Running ssh boro-52 'set -e; for file in /tmp/pahender/*_my_daos.log /tmp/pahender/my_daos.log /tmp/daos_agent.log /tmp/daos_control.log /tmp/test_server.log; do if [ -e $file ]; then scp $file boro-41:/home/pahender/avocado/job-results/latest/daos_logs/${file##*/}-boro-52 && rm -fr $file; fi; done'
Running ssh boro-53 'set -e; for file in /tmp/pahender/*_my_daos.log /tmp/pahender/my_daos.log /tmp/daos_agent.log /tmp/daos_control.log /tmp/test_server.log; do if [ -e $file ]; then scp $file boro-41:/home/pahender/avocado/job-results/latest/daos_logs/${file##*/}-boro-53 && rm -fr $file; fi; done'
Error executing 'ssh boro-52 'set -e; for file in /tmp/pahender/*_my_daos.log /tmp/pahender/my_daos.log /tmp/daos_agent.log /tmp/daos_control.log /tmp/test_server.log; do if [ -e $file ]; then scp $file boro-41:/home/pahender/avocado/job-results/latest/daos_logs/${file##*/}-boro-52 && rm -fr $file; fi; done'':
        Command 'ssh boro-52 'set -e; for file in /tmp/pahender/*_my_daos.log /tmp/pahender/my_daos.log /tmp/daos_agent.log /tmp/daos_control.log /tmp/test_server.log; do if [ -e $file ]; then scp $file boro-41:/home/pahender/avocado/job-results/latest/daos_logs/${file##*/}-boro-52 && rm -fr $file; fi; done'' returned non-zero exit status 1
Renamed /home/pahender/avocado/job-results/job-2019-06-13T12.55-0792328 to /home/pahender/avocado/job-results/job-2019-06-13T12.55-0792328-container_create
```
In this particular case there was an issue copying the `boro-52:/tmp/test_server.log`.  For now this is probably sufficient, but in the future it would be nice to see the actual scp error to get a better understanding of why it failed.

Reply:
So this is the desired effect then yes?  The error propagating up.

What was the cause of the error in copying in this case?  Do we know?

Reply:
Yes, this is the desired effect.  At least we have a notification there was a problem copying at least one file.  As you noted, due to the current setup of `spawn_commands` and `get_output` we do not display the error messages from the failed command output.  If we catch the `subprocess.CalledProcessError` exception for the `subprocess.check_output()` call we can get the output.  If I need another commit I can add this or enable it in a future enhancement.